colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,290 @@
+        return param_scale * rel_step_sz
+
+    @staticmethod
+    def _get_options(param_group, param_shape): 
+        """
+        Determines whether the current param is factored
+        Args:
+            param_group : param group
+            param_shape : Original Shape of param
+            
+        """
+        factored = (len(param_shape) >= 2) 
+        use_first_moment = param_group["beta1"] is not None
+        return factored, use_first_moment
+
+    @staticmethod         
+    def _rms(tensor):
+        return tensor.norm(2) / (tensor.numel() ** 0.5)
+
+    @staticmethod
+    def _approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col):
+        r_factor = (exp_avg_sq_row / exp_avg_sq_row.mean(dim=-1, keepdim=True)).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+    
+    # approx_sq_grad for row parallel weight
+    @staticmethod
+    def _approx_sq_grad_row_parallel(exp_avg_sq_row, exp_avg_sq_col, sq_row_meam):
+        # row_meam = sq_row_meam
+        r_factor = (exp_avg_sq_row / sq_row_meam).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+        
+    @torch.no_grad()
+    def step(self, closure=None):
+        """
+        Performs a single optimization steps
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            loss = closure()
+        """
+        param_groups: Dict
+        {
+            "params":[weight, bias]
+            "lr"
+            "eps"
+            "clip_threshold"
+            "decay_rate"
+            "beta1"
+            "weight_decay"
+            "scale_parameter"
+            "relative_step"
+            "warmup_init"
+        }
+        """
+        # print(f"param_groups {self.param_groups}")
+        for group in self.param_groups:
+            # update weight & bias
+            for p in group["params"]:
+                if p.grad is None:
+                    continue
+                
+                # ==============================
+                # Init Current Param Shape
+                # ==============================
+                p_shape = None  # param shape (not split)
+                param_height = None
+                param_width = None
+                param_height_parallel = None
+                param_width_parallel = None
+                sharding_spec = None
+                if id(p) in self.param_shape.keys():
+                    p_shape = self.param_shape[id(p)]
+                    # if id(p) in self.sharding_spec_dict.keys() and self.sharding_spec_dict[id(p)] != None:
+                    #     sharding_spec = self.sharding_spec_dict[id(p)]
+                    if len(p_shape) >= 2: # factored
+                        param_height = p_shape[0]
```

Comment:
This type of infomation can be stored in class. Getting it in every step takes much time

Reply:
Modified in: https://github.com/hpcaitech/ColossalAI/pull/5484/commits/ce89a849882e845afcc3e99f4a2bdfd9c40f4ee6